src/marqo/tensor_search/throttling/redis_throttle.py

```diff
@@ -0,0 +1,110 @@
+from marqo.tensor_search.enums import ThrottleType
+from marqo.connections import redis_driver
+from marqo.tensor_search.enums import RequestType, EnvVars
+from marqo.tensor_search import utils
+from marqo.tensor_search.tensor_search_logging import get_logger
+from marqo.errors import TooManyRequestsError
+from functools import wraps
+from threading import Thread
+import uuid
+
+# for logging
+import datetime
+import time
+import os
+import logging
+
+logger = get_logger(__name__)
+
+def throttle(request_type: str):
+    """
+    Decorator that checks if a user has exceeded their throttling limits.
+    """
+    def decorator(function):
+        
+        @wraps(function)        # needed to preserve function metadata, or else FastAPI throws a 422.
+        def wrapper(*args, **kwargs):
+
+            # Can be turned off with MARQO_ENABLE_THROTTLING = 'FALSE'
+            if utils.read_env_vars_and_defaults(EnvVars.MARQO_ENABLE_THROTTLING) != "TRUE":
+                return function(*args, **kwargs)
+
+            redis = redis_driver.get_db()  # redis instance
+
+            logger.info(f"Beginning throttling check. API request is of type {request_type}")
+            lua_shas = redis_driver.get_lua_shas()
+
+            # Define maximum thread counts
+            throttling_max_threads = {
+                RequestType.INDEX: utils.read_env_vars_and_defaults(EnvVars.MARQO_MAX_CONCURRENT_INDEX),
+                RequestType.SEARCH: utils.read_env_vars_and_defaults(EnvVars.MARQO_MAX_CONCURRENT_SEARCH) 
+            }
+            
+            set_key = f"set:{request_type}"
+            thread_name = f"thread:{uuid.uuid4()}"
+
+            t0 = time.time()
+
+            # Check current thread count / increment using LUA script
+            try:
+                check_result = redis.evalsha(

```

Comment:
Should we add a timeout? on the rare occasion (perhaps Redis is cooked) that it takes 2 sec+ to respond, for example? 

https://stackoverflow.com/questions/56811414/how-to-set-connection-timeout-in-flask-redis-cache

Reply:
built-in socket_timeout for redis not working. will look into this int he future

Reply:
Create an issue/comment